### PR TITLE
WT-10240 Implement mirrored dynamic tables in workgen

### PR DESCRIPTION
**Pull request by tammybailey**
This PR adds mirror table functionality to workgen for dynamic tables. Tables created by the `Table()` class are not affected. I used the `tools/wt_cmp_uri` script to validate mirrored content (thanks @ddanderson!!)

Changes include:
* Added new members to the `TableRuntime` to indicate if a table is a base table or a mirror.
* Added a boolean to the `app_metadata` indicating if the table is a base or mirror.
* Added the mirror table name to the `app_metadata` if the table is mirrored.
* Updated the dynamic table create and drop to operate on mirror table pairs.
* Moved random table selection out of `op_run` and into a new `op_run_prepare` function so that (non-transaction) mirrored table operations could be wrapped in a transaction.
* Removed the `force` config setting from the `session->drop` calls (see WT-10576) 
* Added a workload option to enable mirroring.
* Added a workload option to specify the suffix for table mirrors.

### CODE REVIEW

**Comment by tammybailey** [comment]
@sulabhM I have this in draft mode while [WT-10576](https://jira.mongodb.org/browse/WT-10576) is under investigation but if you would take a look when you have some time I'd appreciate it :)

**Comment by sulabhM** [comment]
@tammybailey , I had a quick look. Good work on the change, it seems to be in the right direction.

File: bench/workgen/workgen.cpp [hunk 1]
```diff
@@ -397,57 +461,89 @@ WorkloadRunner::start_tables_create(WT_CONNECTION *conn)
...
+
+    // Flag table for deletion.
+    VERBOSE(*_workload, "Flagging pending delete for: " << table_itr->first);
+    std::cerr << "Flagging pending delete for: " << table_itr->first << std::endl;
+    icontext->_dyn_table_runtime[table_itr->second]._pending_delete = true;
+
+    // Add table to the pending delete list.
+    ASSERT(std::find(pending_delete.begin(), pending_delete.end(), table_itr->first) ==
+      pending_delete.end());
+    pending_delete.push_back(table_itr->first);
```
Comment (ddanderson) [inline bench/workgen/workgen.cpp:528]: This code is repeated below, please refactor with a function.

File: bench/workgen/workgen.cpp [hunk 2]
```diff
@@ -1328,37 +1443,84 @@ ThreadRunner::op_get_key_recno(Operation *op, uint64_t range, tint_t tint)
...
+ThreadRunner::op_clear_table(Operation *op)
+{
+    if (op->_random_table && op->is_table_op()) {
+        op->_table._uri = std::string();
+        op->_table._internal->_tint = 0;
+    }
+}
+
+// Generate and set a key and value for the operation managed by this thread.
+// Set the table uri and tint value for a dynamic table operation.
```
Comment (ddanderson) [inline bench/workgen/workgen.cpp:1457]: Looks like this comment applies to the `op_kv_gen` function, not this one?

File: bench/workgen/workgen.cpp [hunk 3]
```diff
@@ -1328,37 +1443,84 @@ ThreadRunner::op_get_key_recno(Operation *op, uint64_t range, tint_t tint)
...
+          (op->_table.options.random_value || _workload->options.random_table_values) ?
+          0 :
+          op->_table.options.value_compressibility;
+        op->kv_gen(this, false, compressibility, recno, _valuebuf);
+    }
+}
+
 int
-ThreadRunner::op_run(Operation *op)
+ThreadRunner::op_run_prepare(Operation *op)
```
Comment (ddanderson) [inline bench/workgen/workgen.cpp:1513]: If I'm not mistaken, there is no transactional "prepare" implied by this function name?  If that's true, I'd suggest renaming it - I was looking for a prepare call.

File: bench/workgen/workgen.cpp [hunk 4]
```diff
@@ -1367,90 +1529,132 @@ ThreadRunner::op_run(Operation *op)
...
+    std::vector<Operation> new_group = {base_op, mirror_op};
+    new_op._group = &new_group;
+    new_op._repeatgroup = 1;
+
+    if (!_in_transaction) {
+        Transaction txn;
+        new_op.transaction = &txn;
+    }
+
+    return op_run(&new_op);
```
Comment (ddanderson) [inline bench/workgen/workgen.cpp:1601]: I love this code!

Except for one thing.  I think when you have:
```
 if (!_in_transaction) {
        Transaction txn;
        new_op.transaction = &txn;
    }
```
txn is on the stack, and when we exit the `}`, isn't its address invalid?  So I think Transction txn must be declared outside the `if` clause (and if it's not used, that's okay).

**Review by ddanderson — APPROVED** [review]
Generally, this looks great.  I think I spotted one bad address reference, and the rest are suggestions or more minor.

I'm out next week, so no need for re-review, but happy to take another look if you finish it today.

File: bench/workgen/workgen.cpp (hunk 4 above)
Comment (tammybailey) [reply bench/workgen/workgen.cpp]: Good catch - thanks! Fixed in [6c72925](https://github.com/wiredtiger/wiredtiger/pull/8798/commits/6c7292568c1e15f405d3e32eec637aca45361cdc).

File: bench/workgen/workgen.cpp (hunk 2 above)
Comment (tammybailey) [reply bench/workgen/workgen.cpp]: Yes it does, I missed it when shuffling code around. Fixed in [74294b2](https://github.com/wiredtiger/wiredtiger/pull/8798/commits/74294b2b7c633043571571c79b4d43a7f71ff7ba).

File: bench/workgen/workgen.cpp (hunk 3 above)
Comment (tammybailey) [reply bench/workgen/workgen.cpp]: Fair point! I changed it to `op_run_setup` in [7dbaa36](https://github.com/wiredtiger/wiredtiger/pull/8798/commits/7dbaa36f79aee9ff5e042633644e2ce5e522c477) but open to suggestions for something better.

File: bench/workgen/workgen.cpp [hunk 5]
```diff
@@ -341,6 +344,48 @@ gen_random_table_name(char *name, workgen_random_state volatile *rand_state)
...
+            return EEXIST;
+    }
+
+    // Create the table.
+    WT_DECL_RET;
+    if ((ret = session->create(session, uri.c_str(), config.c_str())) != 0) {
+        if (ret != EBUSY)
+            THROW("Failed to create table '" << uri << "'.");
+        return ret;
+    }
```
[the remark this answers is not in the document]
Comment (tammybailey) [reply bench/workgen/workgen.cpp:370]: I'm going to call on @sulabhM to get his thoughts as he did the original implementation for this code and spent a fair amount of time on the locking in particular.

[the remark this answers is not in the document]
Comment (sulabhM) [reply bench/workgen/workgen.cpp:370]: @ddanderson , @tammybailey - yes I am keeping the slow operation outside the lock. This is a single thread that does creates, so we don't have to worry about anyone else getting in the way. The first lock to check the tables is the shared lock - since it can co-exist with the operation threads that take the lock on the structure when selecting a table to operate on. The next lock is the exclusive lock that blocks the other shared locks in the meanwhile.

In short, this thread won't race with anyone else creating tables - so we are good. The two part shared lock - exclusive lock is to minimize readers having to wait longer than needed and keeping slow schema operations outside the lock.

File: bench/workgen/workgen.cpp (hunk 3 above)
Comment (ddanderson) [reply bench/workgen/workgen.cpp]: `op_run_setup` sounds good.

File: bench/workgen/workgen.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (ddanderson) [reply bench/workgen/workgen.cpp:370]: Ah, thanks @sulabhM.  That makes sense.

File: bench/workgen/workgen.cpp (hunk 1 above)
Comment (tammybailey) [reply bench/workgen/workgen.cpp]: Done! See [2be2f3b](https://github.com/wiredtiger/wiredtiger/pull/8798/commits/2be2f3b5ca7117dd1299929a57a10dae0a33bea1).

**Comment by ddanderson** [comment]
@tammybailey, all the changes LGTM!

File: bench/workgen/workgen.cpp [hunk 6]
```diff
@@ -366,6 +410,25 @@ WorkloadRunner::start_tables_create(WT_CONNECTION *conn)
...
+    char rand_chars[DYNAMIC_TABLE_LEN];
+
+    // Add app_metadata to the config to indicate the table was created dynamically (can be
+    // selected for random deletion), the name of the table's mirror if mirroring is enabled,
+    // and if this table is a base table or a mirror. We want these settings to persist over
+    // restarts.
+    std::string base_config =
+      "key_format=S,value_format=S,app_metadata=\"" + DYN_TABLE_APP_METADATA;
+    if (_workload->options.mirror_tables) {
+        base_config += "," + MIRROR_TABLE_APP_METADATA;
```
Comment (sulabhM) [inline bench/workgen/workgen.cpp:426]: Is there a + "true" missing here for the mirror table app metadata, or is the intention to fill in the mirror table name here?

File: bench/workgen/workgen.cpp [hunk 7]
```diff
@@ -793,24 +875,51 @@ ContextInternal::create_all(WT_CONNECTION *conn)
...
+            size_t start, end;
+            pos = value.find(MIRROR_TABLE_APP_METADATA);
+            // If the table has a mirror, add the mirror name to the runtime data.
+            if (pos != std::string::npos) {
+                start = value.find_first_of('=', pos);
+                end = value.find_first_of(",\"", start);
+                if (start == std::string::npos || end == std::string::npos) {
+                    THROW("Unable to retrieve mirror name from metadata.");
+                }
+                _dyn_table_runtime[_dyn_tint_last]._mirror =
```
Comment (sulabhM) [inline bench/workgen/workgen.cpp:907]: Does a mirror table also get a mirror table name (with _is_base as false ?

**Review by sulabhM — APPROVED** [review]
I have a couple of questions. Otherwise LGTM.

File: bench/workgen/workgen.cpp (hunk 6 above)
Comment (tammybailey) [reply bench/workgen/workgen.cpp:426]: The intention is to fill in the name of the mirror here.

File: bench/workgen/workgen.cpp (hunk 7 above)
Comment (tammybailey) [reply bench/workgen/workgen.cpp:907]: Yes, the base table gets the name of the mirror and the mirror table gets the name of the base.
